### PR TITLE
Reunite checkpoint and backup core logic

### PR DESCRIPTION
**Pull request by ajkr**
These code paths forked when checkpoint was introduced by copy/pasting the core backup logic. Over time they diverged and bug fixes were sometimes applied to one but not the other (like fix to include all relevant WALs for 2PC), or it required extra effort to fix both (like fix to forge CURRENT file). This diff reunites the code paths by extracting the core logic into a function, CreateCustomCheckpoint(), that is customizable via callbacks to implement both checkpoint and backup.

Related changes:

- flush_before_backup is now forcibly enabled when 2PC is enabled
- Extracted CheckpointImpl class definition into a header file. This is so the function, CreateCustomCheckpoint(), can be called by internal rocksdb code but not exposed to users.
- Implemented more functions in DummyDB/DummyLogFile (in backupable_db_test.cc) that are used by CreateCustomCheckpoint().

### CODE REVIEW

**Comment by siying** [comment]
@ajkr do you plan to rebase?

**Comment by ajkr** [comment]
Yes, maybe will context switch back to this next week, I have some other things to work on now.

**Comment by siying** [comment]
@ajkr makes sense. The next week is the TechDebt week.

**Comment by ajkr** [comment]
ping

**Comment by siying** [comment]
Can you resolve the conflict?

**Comment by ajkr** [comment]
yes as long as you intend to review (otherwise it's kinda pointless to repeatedly resolve merge conflicts).

File: utilities/backupable/backupable_db.cc [hunk 1]
```diff
@@ -742,81 +720,53 @@ Status BackupEngineImpl::CreateNewBackupWithMetadata(
...
+  if (s.ok()) {
+    CheckpointImpl checkpoint(db);
+    uint64_t sequence_number = 0;
+    s = checkpoint.CreateCustomCheckpoint(
+        db->GetDBOptions(),
+        [&](const std::string& src_dirname, const std::string& fname,
+            FileType) {
+          // custom checkpoint will switch to calling copy_file_cb after it sees
+          // NotSupported returned from link_file_cb.
+          return Status::NotSupported();
```
Comment (ajkr) [inline utilities/backupable/backupable_db.cc]: I agree, but this reuses the logic that checkpoint already has for using file copying on Envs that don't support linking. Do you have an alternative idea?

Comment (siying) [inline utilities/backupable/backupable_db.cc]: It's fine then.

**Pull request closed without merging**